packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -171,21 +194,75 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 			invalid: invalid || !!errorSummary,
 		});
 
+		const valueChangeToken = JSON.stringify(
+			value.filter((file) =>
+				fileRejections.every(
+					(rejectedFile) => rejectedFile.file.path !== file.path
+				)
+			)
+		);
+
 		useEffect(() => {
-			setFileRejections(
-				dropzoneFileRejections.map(({ file, errors }) => ({
-					file,
-					errors: errors.map((error) => ({
-						code: error.code,
-						message: getFileRejectionErrorMessage(
-							error,
-							formattedMaxFileSize,
-							acceptedFilesSummary
-						),
-					})),
-				}))
-			);
-		}, [dropzoneFileRejections, formattedMaxFileSize, acceptedFilesSummary]);
+			const isOverMaxFilesLimit =
+				maxFiles && value.length + fileRejections.length > maxFiles;
+
+			if (isOverMaxFilesLimit) {
+				const tooManyFilesRejectionList: Array<RejectedFile> = value
+					.slice(maxFiles)
+					.map((file) => ({
+						file,
+						errors: [tooManyFilesError],
+					}));
+
+				const dropzoneRejections = dropzoneFileRejections.map((rej) => ({
+					...rej,
+					errors: [...rej.errors, tooManyFilesError],
+				}));
+
+				const reformattedDropzoneRejections = reformatDropzoneErrors(
+					dropzoneRejections,
+					maxSize,
+					acceptedFilesSummary
+				);
+
+				const prevRejections = fileRejections.filter((prevFileRej) =>
+					[
+						...tooManyFilesRejectionList,
+						...reformattedDropzoneRejections,
+					].every(
+						(newFileRej) =>
+							JSON.stringify(newFileRej.file) !==
+							JSON.stringify(prevFileRej.file)
+					)
+				);
+				setFileRejections(
+					[
+						...prevRejections,
```

Comment:
prevRejections can be retrieved using `setFileRejections((prevRejections) => { ... })`